lib/ruby_smb/dcerpc/client.rb

```diff
@@ -0,0 +1,521 @@
+            retval = @tcp_socket.write(data[bytes_written..-1])
+            bytes_written += retval
+          end
+
+        rescue IOError, Errno::ECONNABORTED, Errno::ECONNRESET, Errno::EPIPE => e
+          raise Error::CommunicationError, "An error occurred writing to the Socket: #{e.message}"
+        end
+        nil
+      end
+
+      # Receive a packet from the remote host
+      #
+      # @param struct [Class] the structure class to parse the response with
+      # @raise [Error::CommunicationError] if socket-related error occurs
+      def recv_packet(struct)
+        raise Error::CommunicationError, 'Connection has already been closed' if @tcp_socket.closed?
+        if IO.select([@tcp_socket], nil, nil, @read_timeout).nil?
+          raise Error::CommunicationError, "Read timeout expired when reading from the Socket (timeout=#{@read_timeout})"
+        end
+
+        begin
+          response = struct.read(@tcp_socket)
+        rescue IOError
+          raise Error::InvalidPacket, "Error reading the #{struct} response"
+        end
+        unless response.pdu_header.ptype == struct::PTYPE
+          raise Error::InvalidPacket, "Not a #{struct} packet"
+        end
+
+        response
+      rescue Errno::EINVAL, Errno::ECONNABORTED, Errno::ECONNRESET, Errno::EPIPE => e
+        raise Error::CommunicationError, "An error occurred reading from the Socket: #{e.message}"
+      end
+
+      # Process the security context received in a response. It decrypts the
+      # encrypted stub if `:auth_level` is set to anything different than
+      # RPC_C_AUTHN_LEVEL_PKT_PRIVACY. It also checks the packet signature and
+      # raises an InvalidPacket error if it fails. Note that the exception is
+      # disabled by default and can be enabled with the
+      # `:raise_signature_error` option
+      #
+      # @param dcerpc_response [Response] the Response packet
+      #   containing the security context to process
+      # @param opts [Hash] the authenticaiton options: `:auth_type` and
+      #   `:auth_level`. To enable errors when signature check fails, set the
+      #   `:raise_signature_error` option to true
+      # @raise [NotImplementedError] if `:auth_type` is not implemented (yet)
+      # @raise [Error::CommunicationError] if socket-related error occurs
+      def handle_integrity_privacy(dcerpc_response, auth_level:, auth_type:, raise_signature_error: false)
+        decrypted_stub = ''
+        if auth_level == RPC_C_AUTHN_LEVEL_PKT_PRIVACY
+          encrypted_stub = dcerpc_response.stub.to_binary_s + dcerpc_response.auth_pad.to_binary_s
+          case auth_type
+          when RPC_C_AUTHN_WINNT
+            decrypted_stub = @ntlm_client.session.unseal_message(encrypted_stub)
+          when RPC_C_AUTHN_NETLOGON, RPC_C_AUTHN_GSS_NEGOTIATE
+            # TODO
+            raise NotImplementedError
+          else
+            raise ArgumentError, "Unsupported Auth Type: #{auth_type}"
+          end
```

Comment:
I've updated the list of known security providers as defined in [[MS-RPCE] 2.2.1.1.7 Security Providers](https://docs.microsoft.com/en-us/openspecs/windows_protocols/ms-rpce/d4097450-c62f-484b-872f-ddf59a7a0d36) to better define the `NotImplementedError` case. I'm also happy to change `ArgumentError` to `InvalidPacket`, but since `auth_type` comes from an argument passed to the method and not from an actual packet, I would like to make sure it is really what we want.

Reply:
I see your point, from that perspective this makes sense. 👍 I'll mark it as resolved.